content/post/performance-metrics-overview.md

```diff
@@ -0,0 +1,111 @@
+
+Each state entity is managed by a Step, an actor which is responsible for running the state computation and routing output messages downstream, amongst other things. Using Steps allow us to avoid coordination in updating state, but we also need to avoid coordination in our metrics capturing system. To get a full picture of how a state computation is performing, the metrics provided by each Step needs to be able to be stitched together. Here's a diagram to best illustrate the type of aggregation we require across Steps:
+
+![aggregated steps diagram](/images/post/performance-metrics-overview/steps-diagram.png)
+
+A [Sorted Values List](https://blog.wallaroolabs.com/2018/02/missing-slug#using-a-sorted-values-list) would give us the granularity we need to do the aggregations required but would have a much higher performance impact than we want. Another option would be to store specific statistics (mean/median/average) per Step, but the need for aggregation would render these stats entirely useless if we want an accurate depiction of the computation multiple Steps represent. [Theo Schlossnagle](https://twitter.com/postwait) wrote a great blog post explaining why bad math like an average of averages provides useless metrics: ["The Problem with Math: Why Your Monitoring Solution is Wrong"](https://www.circonus.com/2015/02/problem-math/). Existing industry research shows that histograms are excellent solutions for gathering meaningful metrics: ["How NOT to Measure Latency"](https://www.youtube.com/watch?v=lJ8ydIuPFeU) by [Gil Tene](https://twitter.com/giltene) and ["The Uphill Battle for Visibility"](https://www.circonus.com/2016/07/uphill-battle-visibility/) by [Theo Schlossnagle](https://twitter.com/postwait) are just two great examples. We recognized that histograms would work well for us due to the following:
+
+- low cost of both time and space compared to other metric capturing techniques
+- histograms can be easily aggregated, necessary when not all metrics are stored in a single data structure
+- Allows us to answer: is the 99th percentile latency below this time value?
+
+## Minimizing the Metrics Capturing Overhead
+
+We know what metrics information we need to capture, and we knew how we want to capture it. The next step is to determine how much of this workload is required to be handled by Wallaroo. We needed to aggregate histograms to get a complete picture of components but was it worth the resource cost to process this online in Wallaroo? We determined it wasn't. An external system could take these histograms and perform the aggregations needed without impinging on the resources required by Wallaroo. This means freeing up CPU and memory resources that we'd be using if Wallaroo were also responsible for the aggregation of its metrics. Since metrics would have to be offloaded from Wallaroo in some fashion eventually, we decided to do this as early as possible.
```

Comment:
The tenses of `know` and `knew` in the following don't match:

> We know what metrics information we need to capture, and we knew how we want to capture it.

I'm assuming the `knew` should be `know`.

Reply:
There are other tense inconsistencies in the rest of the paragraph also. There is a past tense sentence or two in the middle that is different from the rest.